Run xhr transformers after request interceptors and refactoring

alexanderchr opened the pull request:
* Transformers responsible for modifying the xhr are renamed xhrTransformers to distinguish them from transformers which modify the message (those defined in RequestBuilder and applied in HttpClient).

* A base class RequestMessage is created to contain common functionality of HttpRequestMessage and JSONPRequestMessage

* The xhr is opened and transformed at a later stage, after any request interceptors are run. This allows interceptors to modify the request and closes #60.

Sorry for mashing unrelated commits together, but as two of them are really minor i thought it would save some merging.

EisenbergEffect commented:
@bryanrsmith Can you review this? Also, can you take this into account with respect to the fetch client as well? I know we can't make these apis identical, but it seems reasonable to try to carry concepts and behaviors across the two when possible.

bryanrsmith commented on src/request-message-processor.js:
```diff
@@ -2,22 +2,17 @@ import core from 'core-js';
+function applyXhrTransformers(xhrTransformers, client, processor, message, xhr) {
+  var i, ii;
+  for (i = 0, ii = xhrTransformers.length; i < ii; ++i) {
+      xhrTransformers[i](client, this, message, xhr);
```
Should `this` be `processor`?

bryanrsmith commented on src/request-message-processor.js:
```diff
@@ -84,7 +70,10 @@ export class RequestMessageProcessor {
             this.xhr.abort();
-          } else {
+          }
+          else {
```
It'd be great to match the existing style, and leave the `else` on the same line as the previous closing brace.

bryanrsmith commented on src/request-message.js:
```diff
@@ -0,0 +1,22 @@
+    this.headers = headers || new Headers();
+  }
+
+  fullUrl() {
```
I think the old name of `buildFullUrl` is clearer because the verb clearly indicates what the method does.

bryanrsmith commented on test/http-client.spec.js:
```diff
@@ -535,14 +535,11 @@ describe('http client', () => {
 
       var promise = client.send(new HttpRequestMessage('GET', 'some/cool/path'));
-      promise.catch((response, reason) => {
+      promise.then(() => { done(); }, (response, reason) => {
```
I don't think this is right. It looks like this test will now pass if the promise resolves instead of rejects. We should verify that it doesn't resolve successfully.

bryanrsmith commented on test/http-client.spec.js:
```diff
@@ -787,20 +784,49 @@ describe('http client', () => {
+
+        response(message) {
+          var request = jasmine.Ajax.requests.mostRecent();
+          expect(request.requestHeaders['aheader']).toBe('anothervalue')
```
This test will pass if the client fails to execute the interceptor. I think we could move this assertion into the `.then` continuation on line 812.

bryanrsmith commented on test/http-client.spec.js:
```diff
@@ -787,20 +784,49 @@ describe('http client', () => {
+        });
+    });
+
+    it('should apply xhr transformers before response interceptors are called', (done) => {
```
We should also verify that response interceptors are called before xhr transformers.

alexanderchr replied:
As far as i can see it the xhr transformers has to be called before the request can be made. Are you talking about response transformers?

bryanrsmith replied:
No, I just mean a test that explicitly verifies the change you made to run request interceptors before the xhrTransformers. Something that would break if that ordering were reversed. I suppose the headers test above would break, though, so maybe it's unnecessary.

alexanderchr replied:
Ah. I added a test for that in the processor spec. I think most interceptor tests belong there because it allows testing at a lower level. Do you have any opinion on that?

https://github.com/alexanderchr/http-client/commit/8b51585984190b8d9e625469a87c814b8b80dd49#diff-837ec137ea7cbd265b5ac7a192bda429R176

bryanrsmith replied:
:+1:

bryanrsmith commented on test/request-message-processor.spec.js:
```diff
@@ -54,8 +49,9 @@ describe("Request message processor", () => {
     it("should call xhr.open with the method, full url and ajax set to true", () => {
-      reqProcessor.process(client, message);
-      expect(openSpy).toHaveBeenCalledWith(message.method, message.url, true);
+      reqProcessor.process(client, message).then(() => {
```
This test is now async & should use the `done()` callback.

bryanrsmith commented on test/request-message-processor.spec.js:
```diff
@@ -64,24 +60,16 @@ describe("Request message processor", () => {
-      expect(transformSpy.calls.count()).toBe(2);
+      reqProcessor.xhrTransformers.push(transformSpy);
+      reqProcessor.xhrTransformers.push(transformSpy);
+      reqProcessor.process(client, message).then(() => {
```
Same here.

bryanrsmith commented on test/request-message.spec.js:
```diff
@@ -0,0 +1,25 @@
+    var message;
+    beforeEach(() => {
+      message = new RequestMessage('get', '/anotherurl');
+      message.baseUrl = "example.com";
```
Now that we have a base class, it might be nice to have it initialize its `baseUrl` property to the empty string, so things won't break if we fail to set it.

Also, please use single quotes for strings.

bryanrsmith commented:
Thanks, @alexanderchr! I think this has been bugging a lot of people. Would you mind making a couple cleanup changes?

1. We update the dist/ files at release time, and not with every commit. Could you reset those changes so the commits include only files in src/?
2. I agree that "xhrTransformers" is a clearer name, but that file also includes response transformers that are no longer appropriate under that name.
3. Would you mind updating the commit messages to conform to our [commit message guidelines](https://github.com/DurandalProject/about/blob/master/CONTRIBUTING.md#commit)? Using that format makes it much easier for us to track changes for release notes and such.

Otherwise this looks good to me!

alexanderchr commented:
Thank you very much for the feedback @bryanrsmith. I've fixed most of your comments and updated the commits.

I don't understand your second point though. As far as i can see there are no response transformers in `src/transformers.js` Am i missing something?

bryanrsmith commented on test/http-client.spec.js:
```diff
@@ -534,21 +534,18 @@ describe('http client', () => {
-      // TODO: Enable and fix this test when
-      // https://github.com/jasmine/jasmine-ajax/issues/111
-      // will be fixed
+    fit('should reject when aborted', (done) => {
```
Accidental `fit`? I see a couple minor formatting issues in this test too.

alexanderchr replied:
Accidental indeed. Thanks. I can't see the formatting issues myself, but if you'd show me a perfectly formatted spec I'll gladly put this and all other tests up to that standard.

bryanrsmith commented:
@alexanderchr Ah, my mistake. I was looking at `contentTransformer`, which doesn't mutate the xhr. I think it's close enough, though :-)

alexanderchr commented:
That one bothered me as well. I reasoned that while not strictly an xhr transformer it is something that should also be executed just before the request. Maybe pre-request-transformers would be a better name?

bryanrsmith commented:
@alexanderchr I think it's fine. I just misread the code. Even though it doesn't direct manipulate the XHR, I agree that it's perfectly clear and a reasonable name.

bryanrsmith commented:
@EisenbergEffect Looks good to me! Please verify CLA and merge at your convenience.

EisenbergEffect commented:
Verified. I will merge shortly and this will go out int he release, hopefully today.
